### ISSUE TITLE
Missing validation in HPA

### ISSUE THREAD

**Issue by MaciekPytel** [kubernetes/kubernetes#60511]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
It looks like validation of HPA is insufficient:
- metric name in pods and object metrics should be checked using IsValidPathSegmentName as they're used as part of request to custom metrics API (this is analogous to https://github.com/kubernetes/kubernetes/pull/60433).
- ValidateCrossVersionObjectReference allows for empty apiVersion, leading to non-functional HPA.

**Comment by MaciekPytel** [comment]
/sig-autoscaling

@DirectXMan12

**Comment by DirectXMan12** [comment]
@kubernetes/api-approvers is it fine to update the validation here?  It would reject some previously-accepted HPAs, but those HPAs won't actually work.

**Comment by liggitt** [comment]
> @kubernetes/api-approvers is it fine to update the validation here? It would reject some previously-accepted HPAs, but those HPAs won't actually work.

would it cause problems for the HPA controller trying to update status on the object? would it block deletion of the object (if GC or deletion had to do an update to remove a finalizer before deletion)?

**Comment by DirectXMan12** [comment]
> would it cause problems for the HPA controller trying to update status on the object?

Probably, but that HPA would have been non-functional.

> would it block deletion of the object (if GC or deletion had to do an update to remove a finalizer before deletion)?

Hmm... yeah, yeah, it might.

**Comment by liggitt** [comment]
We also have to ensure that a no-op update is not blocked by tighter valdiation.

> What's the recommended path forward on issues like these, where validation previously let invalid objects through?

We've done ratcheting validation in a few one-off cases, where we always do the extra validation on create, and on update do the extra validation if the old object passed the extra validation (prevents you from making a valid object invalid)

**Comment by mattjmcnaughton** [comment]
@DirectXMan12 @liggitt happy to take a shot at implementing this if no one else is working on it - would be a fun chance to work on part of the code I haven't really dealt with before :)

**Comment by DirectXMan12** [comment]
@mattjmcnaughton go ahead!

/assing @mattjmcnaughton

**Comment by mattjmcnaughton** [comment]
Sweet, I'll get started :)

**Comment by bgrant0607** [comment]
Another caution on tightening validation: be careful not to affect client-side code (e.g., marking fields as required in types.go that are non-optional in validation).

@liggitt It would be useful to document your advice about tightening validation in api_changes.md.

**Comment by liggitt** [comment]
see https://github.com/kubernetes/kubernetes/pull/64907 for a potential mechanism for safely tightening validation (should only be used in cases where existing loose validation allows completely broken objects). if that is merged, that could be used to improve validation here

**Comment by MaciekPytel** [comment]
By now I've seen many people who didn't put apiVersion in object reference, resulting in completely broken HPA objects. So I'm strong +1 for fixing this once the mechanism is available.

**Comment by mattjmcnaughton** [comment]
@DirectXMan12 I apologize I haven't had the time to make progress on this. Please feel free to remove me from the assigned label.

**Comment by DirectXMan12** [comment]
no worries, it happens

**Comment by DirectXMan12** [comment]
go ahead!  We're always glad to have contributions :-).  I think that validation still needs to be fixed.

/assign @Rajat-0

**Comment by pbetkier** [comment]
/remove-lifecycle frozen

nothing happened for almost 5 years, let it get stale if indeed there's no interest anymore

**Comment by lalitc375** [comment]
Let me see how this can be done.

/assign @lalitc375

**Comment by lalitc375** [comment]
IsValidPathSegmentName is already [invoked](https://github.com/kubernetes/kubernetes/blob/285c6113eab3ed21bfa328f6ca99a52bec64a0ed/pkg/apis/autoscaling/validation/validation.go#L358) for metric name in pods and object metrics since https://github.com/kubernetes/kubernetes/commit/a79916fa8443b1920a8c6ca3de502b757ef11878 .

**Comment by lalitc375** [comment]
```go
	// apiVersion is the API version of the referent
	// +optional
	APIVersion string
```

[APIVersion](https://github.com/kubernetes/kubernetes/blob/285c6113eab3ed21bfa328f6ca99a52bec64a0ed/pkg/apis/autoscaling/types.go#L73) is marked optional. Not sure, when it is mandatory. Checking with  sig autoscaling folks. 

cc: @kwiesmueller @laoj2

**Comment by liggitt** [comment]
it looks like it accepts partial values and sometimes parses out group/version and only uses the group portion of the field

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L756-L767

`""` parses to `Group:"", Version:""` with no error, and `Group:""` is valid for the "core" API group. It looks like it _might_ be possible for that field to produce successful autoscaling. If so, we cannot forbid it as an invalid value. This should be double-checked to be sure it is possible to get a working HPA this way.

**Comment by lalitc375** [comment]
Empty("") APIVersion is not supported in HPA for scaleTargetRef field of HPASpec.

```
E0624 22:23:24.787800       1 horizontal.go:278] "Unhandled Error" err="unable to determine resource for scale target reference: no matches for kind \"Deployment\" in group \"\"" logger="UnhandledError"
```

**Comment by liggitt** [comment]
That's true for deployment (which lives in the apps group), but could work properly for replicationcontrollers

**Comment by lalitc375** [comment]
Yes, it is working for replicationcontrollers with empty APIVersion. I think this is the only one. @liggitt  Are you able to think of any other type which can work with empty APIVersion?
